_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+    ofType(getActivityType),
+    mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity')
+      .pipe(
+        map(response => (addActivity(response))),
+        catchError(() => EMPTY)
+      ))
+  ));
+}
+```
+
+Pojawia się tutaj dość dużo zagadnień, zatem spójrzmy na ten fragment kodu od góry do dołu. Najpierw informujemy Angulara, że efekt to tak naprawdę wstrzykiwalna klasa, podobnie jak zwykłe serwisy. W konstruktorze zapewniamy dostęp do serwisu HTTP oraz do zmiennej typu `Actions`. Jest to w gruncie rzeczy serwis udostępniający wszystkie wywołane akcje od momentu ostatniego uruchomienia reducera (czyli dzięki temu damy radę reagować na przychodzące akcje). 
+
+Następnie definiujemy sam efekt:
+* zaglądamy do strumienia akcji 
+```typescript 
+this.actions$.pipe()
+```
+* szukamy tylko akcji określonego typu 
+```typescript
+ofType(getActivityType)
+```
+* reagujemy na każdą z nich poprzez wywołanie żądania HTTP i zaglądamy do strumienia zawierającego odpowiedź z API 
+```typescript 
+mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity').pipe()
+```
+* mapujemy odpowiedź na wywołanie akcji `map(response => (addActivity(response)))`
+```typescript 
+map(response => (addActivity(response)))
+```
+* w przypadku błędu zwracamy pustą wartość Obervable
+```typescript 
+catchError(() => EMPTY)
+```
+
+W wyniku działania tego efektu każde wywołanie akcji `Get Activity` przy poprawnej odpowiedzi z API wywoła akcję `Add Activity` z otrzymaną czynnością.
+
+### Reducer
+W reducerach zazwyczaj znajduje się najwięcej logiki ze wszystkich komponentów NgRx. To tutaj trzeba zdecydować jak dodawać do store kolejne dane. Warto pamiętać, że reducery to **czyste funkcje**, które przyjmują jako parametr akcję i obecny stan aplikacji, a następnie zwracają nowy stan, dzięki czemu zachowujemy jego niemutowalność. W naszym przypadku wygląda to tak:
+
+```typescript
+const initialState: ActivityItemsState = {activities: []};
+
+export const activityReducer = createReducer(
+    initialState,
+    on(addActivity, (state, payload) => ({
+        ...state,
+        activities: state.activities.concat({activity: payload.activity, participants: payload.participants} as ActivityItemModel)
+    }))
+);
+```
+
+W pierwszej linii definiujemy jakiego stanu początkowego store (lub - jak ma to miejsce powyżej - jego wycinka) się spodziewamy. W naszym przypadku będzie to pusta tablica. Następnie wywołujemy funkcję `createReducer`, która w pierwszym parametrze przyjmuje wspomniany stan początkowy, a w drugim (i każdym kolejnym) - definicję odpowiedzi na przybyłą akcję. W tym przypadku reducer reaguje na akcję `Add Activity` poprzez dodanie do dotychczasowej tablicy `activities` nowego elementu i zwrócenie całości jako wynik wywołania funkcji anonimowej. Reducer jest częścią całej architektury, którą można w bardzo łatwy sposób przetestować i zapewnić sobie w ten sposób pewność jego poprawnego działania.
```

Comment:
Stanu początkowego się nie spodziewamy, mówimy jak go zainicjalizować. Bo nawet w sytuacji, w której odtwarzalibyśmy store to i tak byłby stan poprzedni, a nie początkowy :)

Reply:
Ad 1. - ok
Ad 2. - zapożyczymy
Ad 3. - oczywiście, że nie musi. Ale ja takiej użyłem i dlatego napisałem, że jest anonimowa